rust/template/distributed_datalog/src/instantiate.rs

```diff
@@ -0,0 +1,539 @@
+/// Add file sinks to the given server object, as per the node
+/// configuration.
+fn add_file_sinks<P>(server: &mut DDlogServer<P>, node_cfg: &NodeCfg) -> Result<(), String>
+where
+    P: Send + DDlog + 'static,
+    P::Convert: Send,
+{
+    deduce_sinks_or_sources(node_cfg, true)
+        .iter()
+        .try_for_each(|(path, rel_ids)| {
+            let file = File::create(path)
+                .map_err(|e| format!("failed to create file {}: {}", path.display(), e))?;
+            let sink = FileSink::<P::Convert>::new(file);
+
+            let mut stream = server.add_stream(rel_ids.clone());
+            stream.subscribe(Box::new(sink)).map_err(|_| {
+                format!(
+                    "failed to subscribe file sink {} to DDlogServer",
+                    path.display()
+                )
+            })?;
+            Ok(())
+        })
+}
+
+fn add_file_sources<P>(
+    txnmux: &mut TxnMux<Update<P::Value>, String>,
+    node_cfg: &NodeCfg,
+) -> Result<(), String>
+where
+    P: DDlog + 'static,
+    P::Convert: Send,
+{
+    deduce_sinks_or_sources(node_cfg, false)
+        .iter()
+        .try_for_each(|(path, _rel_ids)| {
+            let source = FileSource::<P::Convert, _>::new(path);
+            txnmux
+                .add_observable(Box::new(source))
+                .map_err(|_| format!("failed to add file source {} to TxnMux", path.display()))?;
+            Ok(())
+        })
+}
+
+/// Realize the given configuration locally.
+// TODO: Right now this function assumes a pristine state (i.e., nothing
+//       had been created previously), however we really would want to
+//       transition from a previously created state (which happens to be
+//       "empty" initially) to the given one.
+fn realize<P>(
+    addr: &Addr,
+    node_cfg: &NodeCfg,
+    outputs: Outputs,
+) -> Result<Realization<P::Value>, String>
+where
+    P: Send + DDlog + 'static,
+    P::Convert: Send,
+{
+    let mut server = create_server::<P>(&node_cfg)?;
+    add_tcp_senders(&mut server, outputs)?;
+    add_file_sinks(&mut server, node_cfg)?;
+
+    let mut txnmux = create_txn_mux(server)?;
+    add_tcp_receiver(&mut txnmux, addr)?;
+    add_file_sources::<P>(&mut txnmux, node_cfg)?;
+
+    Ok(Realization { txnmux })
+}
+
+/// An object representing a realized configuration.
+///
+/// Right now all clients can do with object of this type is dropping
+/// them to tear everything down.
+#[derive(Debug)]
+pub struct Realization<V>
+where
+    V: Debug + Send,
+{
+    /// The transaction multiplexer everything is registered to.
+    txnmux: TxnMux<Update<V>, String>,
+}
+
+/// Instantiate a configuration on a particular node under the given
+/// assignment.
+pub fn instantiate<P>(
+    sys_cfg: SysCfg,
+    addr: &Addr,
+    assignment: &Assignment,
+) -> Result<Vec<Realization<P::Value>>, String>
+where
+    P: Send + DDlog + 'static,
+    P::Convert: Send,
+{
+    assignment
+        .iter()
+        .filter_map(|(uuid, assigned_addr)| {
```

Comment:
Wouldn't it be more natural for this function to take node id instead of address as an argument and then this `filter_map` becomes just a map lookup.

Reply:
Hm. Not the way I see it. If we knew the node id we pretty much wouldn't need the function to begin with. The way I envisioned it is that an assignment can actually map multiple `Node`s to one `Addr` (i.e., one member). A member "knows" its own address. So here we may actually end up "realizing" multiple `NodeCfg`s.

Reply:
We discussed this in person. The comment I made is misleading and no longer true for the current version of the code. Previously, I was assuming the port not to be part of the configuration and then multiple nodes could theoretically have shared the same member.
With the port being in the configuration the addresses would be different (because the port is). I'll see how that simplifies things.